pkg/trtl/mock/counterparty.go

```diff
@@ -0,0 +1,361 @@
+	"github.com/trisacrypto/directory/pkg/trtl/peers/v1"
+	"github.com/trisacrypto/directory/pkg/utils/bufconn"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+const (
+	bufSize     = 1024 * 1024
+	GetRPC      = "trtl.v1.Get"
+	PutRPC      = "trtl.v1.Put"
+	DeleteRPC   = "trtl.v1.Delete"
+	IterRPC     = "trtl.v1.Iter"
+	BatchRPC    = "trtl.v1.Batch"
+	CursorRPC   = "trtl.v1.Cursor"
+	SyncRPC     = "trtl.v1.Sync"
+	StatusRPC   = "trtl.v1.Status"
+	GetPeersRPC = "trtl.peers.v1.PeerManagement.GetPeers"
+	AddPeersRPC = "trtl.peers.v1.PeerManagement.AddPeers"
+	RmPeersRPC  = "trtl.peers.v1.PeerManagement.RmPeers"
+	GossipRPC   = "honu.replica.v1.Replication/Gossip"
+)
+
+// New creates a new mock RemoteTrtl. If bufnet is nil, one is created for the user.
+// The gRPC server defaults to insecure which can be overridden by passing in
+// ServerOptions with configured TLS.
+func New(bufnet *bufconn.GRPCListener, opts ...grpc.ServerOption) *RemoteTrtl {
+	if bufnet == nil {
+		bufnet = bufconn.New(bufSize, "")
+	}
+
+	if len(opts) == 0 {
+		opts = make([]grpc.ServerOption, 0)
+		opts = append(opts, grpc.Creds(insecure.NewCredentials()))
+	}
+
+	remote := &RemoteTrtl{
+		bufnet: bufnet,
+		srv:    grpc.NewServer(opts...),
+		Calls:  make(map[string]int),
+	}
+
+	pb.RegisterTrtlServer(remote.srv, remote)
+	peers.RegisterPeerManagementServer(remote.srv, remote)
+	go remote.srv.Serve(remote.bufnet.Listener)
+	return remote
+}
+
+// RemoteTrtl implements a mock gRPC server for testing client connections to other
+// trtl peers. The desired response of the remote peer can be set by external callers
+// using the OnRPC functions or the WithFixture or WithError functions. The Calls map
+// can be used to count the number of times the remote peer RPC was called.
+type RemoteTrtl struct {
+	pb.UnimplementedTrtlServer
+	peers.UnimplementedPeerManagementServer
+	replica.UnimplementedReplicationServer
+	bufnet     *bufconn.GRPCListener
+	srv        *grpc.Server
+	Calls      map[string]int
+	OnGet      func(context.Context, *pb.GetRequest) (*pb.GetReply, error)
+	OnPut      func(context.Context, *pb.PutRequest) (*pb.PutReply, error)
+	OnDelete   func(context.Context, *pb.DeleteRequest) (*pb.DeleteReply, error)
+	OnIter     func(context.Context, *pb.IterRequest) (*pb.IterReply, error)
+	OnBatch    func(pb.Trtl_BatchServer) error
+	OnCursor   func(*pb.CursorRequest, pb.Trtl_CursorServer) error
+	OnSync     func(pb.Trtl_SyncServer) error
+	OnStatus   func(context.Context, *pb.HealthCheck) (*pb.ServerStatus, error)
+	OnGetPeers func(context.Context, *peers.PeersFilter) (*peers.PeersList, error)
+	OnAddPeers func(context.Context, *peers.Peer) (*peers.PeersStatus, error)
+	OnRmPeers  func(context.Context, *peers.Peer) (*peers.PeersStatus, error)
+	OnGossip   func(replica.Replication_GossipServer) error
```

Comment:
I know this was a bit of work - thank you so much for adding all of these!

Reply:
Thankfully GitHub Copilot made this a bit easier!